kyuubi-server/src/main/scala/org/apache/kyuubi/server/metadata/MetadataManager.scala

```diff
@@ -0,0 +1,293 @@
+      size,
+      true).map(buildBatch)
+  }
+
+  def getBatchesRecoveryMetadata(
+      state: String,
+      kyuubiInstance: String,
+      from: Int,
+      size: Int): Seq[Metadata] = {
+    _metadataStore.getMetadataList(
+      SessionType.BATCH,
+      null,
+      null,
+      state,
+      kyuubiInstance,
+      0,
+      0,
+      from,
+      size,
+      false)
+  }
+
+  def updateMetadata(metadata: Metadata, retryOnError: Boolean = true): Unit = {
+    try {
+      _metadataStore.updateMetadata(metadata)
+    } catch {
+      case e: Throwable if retryOnError =>
+        error(s"Error updating metadata for session ${metadata.identifier}", e)
+        addMetadataRetryRequest(UpdateMetadata(metadata))
+    }
+  }
+
+  def cleanupMetadataById(batchId: String): Unit = {
+    _metadataStore.cleanupMetadataByIdentifier(batchId)
+  }
+
+  private def buildBatch(batchMetadata: Metadata): Batch = {
+    val batchAppInfo = Map(
+      APP_ID_KEY -> Option(batchMetadata.engineId),
+      APP_NAME_KEY -> Option(batchMetadata.engineName),
+      APP_STATE_KEY -> Option(batchMetadata.engineState),
+      APP_URL_KEY -> Option(batchMetadata.engineUrl),
+      APP_ERROR_KEY -> batchMetadata.engineError)
+      .filter(_._2.isDefined)
+      .map(info => (info._1, info._2.get))
+
+    new Batch(
+      batchMetadata.identifier,
+      batchMetadata.username,
+      batchMetadata.engineType,
+      batchMetadata.requestName,
+      batchAppInfo.asJava,
+      batchMetadata.kyuubiInstance,
+      batchMetadata.state,
+      batchMetadata.createTime,
+      batchMetadata.endTime)
+  }
+
+  private def startMetadataCleaner(): Unit = {
+    val cleanerEnabled = conf.get(KyuubiConf.METADATA_CLEANER_ENABLED)
+    val stateMaxAge = conf.get(METADATA_MAX_AGE)
+
+    if (cleanerEnabled) {
+      val interval = conf.get(KyuubiConf.METADATA_CLEANER_INTERVAL)
+      val cleanerTask: Runnable = () => {
+        try {
+          _metadataStore.cleanupMetadataByAge(stateMaxAge)
+        } catch {
+          case e: Throwable => error("Error cleaning up the metadata by age", e)
+        }
+      }
+
+      metadataCleaner.scheduleWithFixedDelay(
+        cleanerTask,
+        interval,
+        interval,
+        TimeUnit.MILLISECONDS)
+    }
+  }
+
+  def addMetadataRetryRequest(request: MetadataRequest): Unit = {
+    if (identifierRequestsRetryRefs.size() > maxMetadataRequestsRetryRefs) {
+      throw new KyuubiException(
+        "The number of metadata requests retry instances exceeds the limitation:" +
+          maxMetadataRequestsRetryRefs)
+    }
+    val identifier = request.metadata.identifier
+    val ref = identifierRequestsRetryRefs.computeIfAbsent(
+      identifier,
+      identifier => {
+        val ref = new MetadataRequestsRetryRef
+        debug(s"Created MetadataRequestsRetryRef for session $identifier.")
+        ref
+      })
+    ref.addRetryingMetadataRequest(request)
+    identifierRequestsRetryRefs.putIfAbsent(identifier, ref)
```

Comment:
An asynchronous retry looks dangerous IMO, we can NOT ensure that the metadata of batch job persists into RDMS eventually.

I roughly went through the discussion, the original reason was just "to unblock the main thread"? I think it's not much important compared to the risk of lost data.

I would like to change the retry logic from asynchronous to synchronous.

Reply:
It Will wait the retrying metadata request finished before the batch closed.

Reply:
It is not acceptable that the kyuubi service is unavailable on MySQL issue.

Reply:
Maybe you can add audit log for metadata requests，likes rest audit log.